### ISSUE TITLE
units: Make the crate conform to the API guidelines

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3632]
# Rust API Guidelines Checklist

- **Naming** *(crate aligns with Rust naming conventions)*
  - [ ] Casing conforms to RFC 430 ([C-CASE])
  - [ ] Ad-hoc conversions follow `as_`, `to_`, `into_` conventions ([C-CONV])
  - [ ] Getter names follow Rust convention ([C-GETTER])
  - [ ] Methods on collections that produce iterators follow `iter`, `iter_mut`, `into_iter` ([C-ITER])
  - [ ] Iterator type names match the methods that produce them ([C-ITER-TY])
  - [ ] Feature names are free of placeholder words ([C-FEATURE])
  - [ ] Names use a consistent word order ([C-WORD-ORDER])
- **Interoperability** *(crate interacts nicely with other library functionality)*
  - [ ] Types eagerly implement common traits ([C-COMMON-TRAITS])
    - `Copy`, `Clone`, `Eq`, `PartialEq`, `Ord`, `PartialOrd`, `Hash`, `Debug`,
      `Display`, `Default`
  - [ ] Conversions use the standard traits `From`, `AsRef`, `AsMut` ([C-CONV-TRAITS])
  - [ ] Collections implement `FromIterator` and `Extend` ([C-COLLECT])
  - [ ] Data structures implement Serde's `Serialize`, `Deserialize` ([C-SERDE])
  - [ ] Types are `Send` and `Sync` where possible ([C-SEND-SYNC])
  - [ ] Error types are meaningful and well-behaved ([C-GOOD-ERR])
  - [ ] Binary number types provide `Hex`, `Octal`, `Binary` formatting ([C-NUM-FMT])
  - [ ] Generic reader/writer functions take `R: Read` and `W: Write` by value ([C-RW-VALUE])
- **Macros** *(crate presents well-behaved macros)*
  - [ ] Input syntax is evocative of the output ([C-EVOCATIVE])
  - [ ] Macros compose well with attributes ([C-MACRO-ATTR])
  - [ ] Item macros work anywhere that items are allowed ([C-ANYWHERE])
  - [ ] Item macros support visibility specifiers ([C-MACRO-VIS])
  - [ ] Type fragments are flexible ([C-MACRO-TY])
- **Documentation** *(crate is abundantly documented)*
  - [ ] Crate level docs are thorough and include examples ([C-CRATE-DOC])
  - [ ] All items have a rustdoc example ([C-EXAMPLE])
  - [ ] Examples use `?`, not `try!`, not `unwrap` ([C-QUESTION-MARK])
  - [ ] Function docs include error, panic, and safety considerations ([C-FAILURE])
  - [ ] Prose contains hyperlinks to relevant things ([C-LINK])
  - [ ] Cargo.toml includes all common metadata ([C-METADATA])
    - authors, description, license, homepage, documentation, repository,
      keywords, categories
  - [ ] Release notes document all significant changes ([C-RELNOTES])
  - [ ] Rustdoc does not show unhelpful implementation details ([C-HIDDEN])
- **Predictability** *(crate enables legible code that acts how it looks)*
  - [ ] Smart pointers do not add inherent methods ([C-SMART-PTR])
  - [ ] Conversions live on the most specific type involved ([C-CONV-SPECIFIC])
  - [ ] Functions with a clear receiver are methods ([C-METHOD])
  - [ ] Functions do not take out-parameters ([C-NO-OUT])
  - [ ] Operator overloads are unsurprising ([C-OVERLOAD])
  - [ ] Only smart pointers implement `Deref` and `DerefMut` ([C-DEREF])
  - [ ] Constructors are static, inherent methods ([C-CTOR])
- **Flexibility** *(crate supports diverse real-world use cases)*
  - [ ] Functions expose intermediate results to avoid duplicate work ([C-INTERMEDIATE])
  - [ ] Caller decides where to copy and place data ([C-CALLER-CONTROL])
  - [ ] Functions minimize assumptions about parameters by using generics ([C-GENERIC])
  - [ ] Traits are object-safe if they may be useful as a trait object ([C-OBJECT])
- **Type safety** *(crate leverages the type system effectively)*
  - [ ] Newtypes provide static distinctions ([C-NEWTYPE])
  - [ ] Arguments convey meaning through types, not `bool` or `Option` ([C-CUSTOM-TYPE])
  - [ ] Types for a set of flags are `bitflags`, not enums ([C-BITFLAG])
  - [ ] Builders enable construction of complex values ([C-BUILDER])
- **Dependability** *(crate is unlikely to do the wrong thing)*
  - [ ] Functions validate their arguments ([C-VALIDATE])
  - [ ] Destructors never fail ([C-DTOR-FAIL])
  - [ ] Destructors that may block have alternatives ([C-DTOR-BLOCK])
- **Debuggability** *(crate is conducive to easy debugging)*
  - [ ] All public types implement `Debug` ([C-DEBUG])
  - [ ] `Debug` representation is never empty ([C-DEBUG-NONEMPTY])
- **Future proofing** *(crate is free to improve without breaking users' code)*
  - [ ] Sealed traits protect against downstream implementations ([C-SEALED])
  - [ ] Structs have private fields ([C-STRUCT-PRIVATE])
  - [ ] Newtypes encapsulate implementation details ([C-NEWTYPE-HIDE])
  - [ ] Data structures do not duplicate derived trait bounds ([C-STRUCT-BOUNDS])
- **Necessities** *(to whom they matter, they really matter)*
  - [ ] Public dependencies of a stable crate are stable ([C-STABLE])
  - [ ] Crate and its dependencies have a permissive license ([C-PERMISSIVE])

[C-CASE]: https://rust-lang.github.io/api-guidelines/naming.html#c-case
[C-CONV]: https://rust-lang.github.io/api-guidelines/naming.html#c-conv
[C-GETTER]: https://rust-lang.github.io/api-guidelines/naming.html#c-getter
[C-ITER]: https://rust-lang.github.io/api-guidelines/naming.html#c-iter
[C-ITER-TY]: https://rust-lang.github.io/api-guidelines/naming.html#c-iter-ty
[C-FEATURE]: https://rust-lang.github.io/api-guidelines/naming.html#c-feature
[C-WORD-ORDER]: https://rust-lang.github.io/api-guidelines/naming.html#c-word-order

[C-COMMON-TRAITS]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-common-traits
[C-CONV-TRAITS]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-conv-traits
[C-COLLECT]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-collect
[C-SERDE]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-serde
[C-SEND-SYNC]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-send-sync
[C-GOOD-ERR]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-good-err
[C-NUM-FMT]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-num-fmt
[C-RW-VALUE]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-rw-value

[C-EVOCATIVE]: https://rust-lang.github.io/api-guidelines/macros.html#c-evocative
[C-MACRO-ATTR]: https://rust-lang.github.io/api-guidelines/macros.html#c-macro-attr
[C-ANYWHERE]: https://rust-lang.github.io/api-guidelines/macros.html#c-anywhere
[C-MACRO-VIS]: https://rust-lang.github.io/api-guidelines/macros.html#c-macro-vis
[C-MACRO-TY]: https://rust-lang.github.io/api-guidelines/macros.html#c-macro-ty

[C-CRATE-DOC]: https://rust-lang.github.io/api-guidelines/documentation.html#c-crate-doc
[C-EXAMPLE]: https://rust-lang.github.io/api-guidelines/documentation.html#c-example
[C-QUESTION-MARK]: https://rust-lang.github.io/api-guidelines/documentation.html#c-question-mark
[C-FAILURE]: https://rust-lang.github.io/api-guidelines/documentation.html#c-failure
[C-LINK]: https://rust-lang.github.io/api-guidelines/documentation.html#c-link
[C-METADATA]: https://rust-lang.github.io/api-guidelines/documentation.html#c-metadata
[C-HTML-ROOT]: https://rust-lang.github.io/api-guidelines/documentation.html#c-html-root
[C-RELNOTES]: https://rust-lang.github.io/api-guidelines/documentation.html#c-relnotes
[C-HIDDEN]: https://rust-lang.github.io/api-guidelines/documentation.html#c-hidden

[C-SMART-PTR]: https://rust-lang.github.io/api-guidelines/predictability.html#c-smart-ptr
[C-CONV-SPECIFIC]: https://rust-lang.github.io/api-guidelines/predictability.html#c-conv-specific
[C-METHOD]: https://rust-lang.github.io/api-guidelines/predictability.html#c-method
[C-NO-OUT]: https://rust-lang.github.io/api-guidelines/predictability.html#c-no-out
[C-OVERLOAD]: https://rust-lang.github.io/api-guidelines/predictability.html#c-overload
[C-DEREF]: https://rust-lang.github.io/api-guidelines/predictability.html#c-deref
[C-CTOR]: https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor

[C-INTERMEDIATE]: https://rust-lang.github.io/api-guidelines/flexibility.html#c-intermediate
[C-CALLER-CONTROL]: https://rust-lang.github.io/api-guidelines/flexibility.html#c-caller-control
[C-GENERIC]: https://rust-lang.github.io/api-guidelines/flexibility.html#c-generic
[C-OBJECT]: https://rust-lang.github.io/api-guidelines/flexibility.html#c-object

[C-NEWTYPE]: https://rust-lang.github.io/api-guidelines/type-safety.html#c-newtype
[C-CUSTOM-TYPE]: https://rust-lang.github.io/api-guidelines/type-safety.html#c-custom-type
[C-BITFLAG]: https://rust-lang.github.io/api-guidelines/type-safety.html#c-bitflag
[C-BUILDER]: https://rust-lang.github.io/api-guidelines/type-safety.html#c-builder

[C-VALIDATE]: https://rust-lang.github.io/api-guidelines/dependability.html#c-validate
[C-DTOR-FAIL]: https://rust-lang.github.io/api-guidelines/dependability.html#c-dtor-fail
[C-DTOR-BLOCK]: https://rust-lang.github.io/api-guidelines/dependability.html#c-dtor-block

[C-DEBUG]: https://rust-lang.github.io/api-guidelines/debuggability.html#c-debug
[C-DEBUG-NONEMPTY]: https://rust-lang.github.io/api-guidelines/debuggability.html#c-debug-nonempty

[C-SEALED]: https://rust-lang.github.io/api-guidelines/future-proofing.html#c-sealed
[C-STRUCT-PRIVATE]: https://rust-lang.github.io/api-guidelines/future-proofing.html#c-struct-private
[C-NEWTYPE-HIDE]: https://rust-lang.github.io/api-guidelines/future-proofing.html#c-newtype-hide
[C-STRUCT-BOUNDS]: https://rust-lang.github.io/api-guidelines/future-proofing.html#c-struct-bounds

[C-STABLE]: https://rust-lang.github.io/api-guidelines/necessities.html#c-stable
[C-PERMISSIVE]: https://rust-lang.github.io/api-guidelines/necessities.html#c-permissive

**Comment by tcharding** [comment]
C-DEREF: Done, checked with `git grep Deref` - no results.

**Comment by tcharding** [comment]
C-FEATURE: Done, the only features are `default`, `std`, and `alloc`.

**Comment by tcharding** [comment]
Done in #3639:
- C-DEBUG
- C-DEBUG-NONEMPT
-

**Comment by tcharding** [comment]
Done in #3690: C-COMMON-TRAITS (along with `Display` done in #3659)

**Comment by tcharding** [comment]
Scratched out items are checked off over here: https://github.com/rust-bitcoin/rust-bitcoin/issues/1649

**Comment by jamillambert** [comment]
Checked - Documentation examples ([C-QUESTION-MARK](https://rust-lang.github.io/api-guidelines/documentation.html#c-question-mark)). Can be checked off, there are no `try!` or `unwrap`.

**Comment by tcharding** [comment]
C-CONV: Done, manually audited ever 'pub fn' (excluding `signed` because they are the same as `unsigned`).

**Comment by tcharding** [comment]
C-ITER: Done, there are no iterators in `units` (excluding implementations of `iter::Sum`).

**Comment by tcharding** [comment]
Updated ticks in issue description up to here.

**Comment by jamillambert** [comment]
C-WORD-ORDER checked manually by searching for `let`, `fn`, `struct` and `enum` (and looking at enum contents).  All good.

**Comment by tcharding** [comment]
I used the `api` script to double check as well.

**Comment by tcharding** [comment]
C-CONV-TRAITS: Done
- All types that are `Clone` are also `Copy` so we don't need `AsRef` or `AsMut`
- We have what looks like sane imples of `From` and `TryFrom` for all types.

**Comment by tcharding** [comment]
C-COLLECT: Done - there are no collections in `units`.

**Comment by tcharding** [comment]
C-SEND-SYNC: Done in #3706

**Comment by tcharding** [comment]
C-GOOD-ERR: Done - I'll eat my hat if this one is wrong. I did add a test for `Send` and `Sync` in #3706 though.

**Comment by tcharding** [comment]
C-NUM-FMT: Done, `units` does not have any binary types.

**Comment by tcharding** [comment]
C-RW-VALUE: Done. No function in `units` takes a `Read` or `Write` trait object by value.

**Comment by tcharding** [comment]
C-MACRO-VIS: Done, unnecessary because we only have two public macros (plus 2 helper macros) and both add impls not types.

**Comment by tcharding** [comment]
Probably want to tackle all the other macro items as part of #3707

**Comment by tcharding** [comment]
C-STABLE : 

- `internals` - broken (see #3708)
- `serde` - good, v1
- `arbitrayr` good, v1

**Comment by tcharding** [comment]
C-STRUCT-PRIVTE: Done, all non-error structs have private fields except `BlockHeight` and `BlockInterval`  which are public by design.

**Comment by tcharding** [comment]
C-NEWTYPE-HIDE: Done, trivial. The types in this crate _are_ newtypes.

**Comment by tcharding** [comment]
C-STRUCT-BOUNDS: Done, there are no generic data structures in `units`.

**Comment by tcharding** [comment]
C-CUSTOM-TYPE: Done, the only functions that take std types are constructors or conversion functions.

**Comment by tcharding** [comment]
C-BITFLAG: Done, trivial. There are no bitflags in `units`.

**Comment by tcharding** [comment]
C-BUILDER: Unnecessary, `units` types are basically just simple wrapper types. Builder pattern does not apply.

**Comment by tcharding** [comment]
Marked checkmarks to here.

**Comment by tcharding** [comment]
C-INTERMEDIATE: Done, manually audited every function in the crate.

**Comment by tcharding** [comment]
C-CALLER-CONTROL: Done, all the "main" types in this crate are `Copy` (and there are no calls to `clone` in the crate).

**Comment by tcharding** [comment]
C-GENERIC: Done, manually audited all functions. This one is a bit fuzzy but basically there are no functions that are not getters, constructors, or converters. There are no collections in the crate or in function parameters.

**Comment by tcharding** [comment]
C-OBJECT: Done, all traits in the `units` crate are private and not used as trait objects.

**Comment by tcharding** [comment]
C-VALIDATE: Done, audited all functions. Trivial to confirm because basically every functions is a converter, constructor, or getter.

**Comment by tcharding** [comment]
C-DTOR-FAIL and C-DTOR-BLOCK

Because we only use stdlib types in this crate and there are not custom imlps of `Drop` I believe both of these can be considered done.

**Comment by tcharding** [comment]
C-CONST Done (did #3726 as part of it) but we can still check it off.

**Comment by tcharding** [comment]
Predictability:

- C-SMART-PTR: There are no smart pointers in the crate
- C-CONV-SPECIFIC: Types include `to_`, `from_` etc from less specific (stdlib) types
- C-METHOD: Used `gg -P '^pub fn .*?\('` and checked the output
- C-CTOR: I'm confident we have contrustors correct, I grepped about a bit and also did #3726

**Comment by jamillambert** [comment]
I have started to look at all the points in Documentation.  I'll update this comment when done.

**Comment by tcharding** [comment]
Sweeeeet, the examples on is not trivial. Having a "why would you do this" example as mentioned in the API guide I mean. Holla at me if you want a hand with that one.

**Comment by tcharding** [comment]
@jamillambert note that in contradition to rustdocs rules we copy stdlib for ops functions and don't use a `Returns` section. So this is correct

```rust
    /// Checked division.
    ///
    /// Computes `self / rhs` returning [`None`] if `rhs == 0`.
```

**Comment by jamillambert** [comment]
> Sweeeeet, the examples on is not trivial. Having a "why would you do this" example as mentioned in the API guide I mean. Holla at me if you want a hand with that one.

Not trivial indeed.  Trying to work out an example showing why and not just using it takes some thinking.  I did a search in `units` and there are 70x `pub fn` 63x `pub const fn` but only 8x `# Examples`.  I think a lot are variations on a theme and don't need examples, but still.  

I'll knock up a PR just adding examples to make sure I am on the right track and not just wasting my time.

**Comment by tcharding** [comment]
Yeah because `units` is almost all made up of getters, setters, and conversion functions I'm not sure how much value the examples add. @apoelstra do you have an opinion on this?

**Comment by tcharding** [comment]
Hey @jamillambert lets put this on hold until #3794 gets done.

**Comment by tcharding** [comment]
The only macros are the ones in `parse`. There is a good chance that they are either made private (by duplicating across crates) or pulled out into a separate crate (see #3817). Either way they will cease to appear in the public API of this crate so all 5 bullet points under the macros heading can be ommited (checked off as unnecessary).

**Comment by tcharding** [comment]
We will likely need to go over this checklist again (*sigh*) after #4007 and #3794 merge.

**Comment by Kixunil** [comment]
Note that I've added an extra item.

**Comment by tcharding** [comment]
Docs section

- C-FAILURE is checked by the linter because we enable pedantic lints.
- C-RELNOTE: All crates have CHANGELOG.md
- C-METADATA: All our crates have this done already.

I went over the whole crate and apart from `parse` (#4176) and `amount` I rekon the docs are good to go.

`amount` will need another audit after the MAX_MONEY stuff settles

**Comment by tcharding** [comment]
### Documentation section

- C-METADATA: #4340
- C-CRATE-DOC: #4341
- C-EXAMPLE: IMO we have enough examples. Most of the methods in this crate are simply getters or setters or math ops. Anything more complicated is already well documented. Note that we have: #4318 for documenting `NumOpResult`.
- C-FAILURE: As above, I flicked through the whole crate and it LGTM.
- C-RELNOTES: We have these.
- C-HIDDEN: There is not much in the docs, I do not believe anything extra appears that should be hidden.

TODO:

- Go over the rendered docs for the whole crate
- Go over `amount` once all other dev is done.

**Comment by jamillambert** [comment]
I have gone over the rendered docs for the whole crate and made a few changes in #4343

**Comment by tcharding** [comment]
It may be worth going over this again. We are trying to release this week already but no point rushing and getting egg on our faces. I would be willing to put this work at the top of my list and we can smash it out @mpbagot? Any one else on the team willing to throw some effort into this too (@jamillambert @nyonson @satsfy)?

**Comment by tcharding** [comment]
`Macros` section no need to check. There are no public macros in `units`.

**Comment by tcharding** [comment]
`Necessities` section done. All deps are post-1.0

**Comment by tcharding** [comment]
`Future proofing` section

- C-SEALED: there is one trait and it is sealed
- C-STRUCT-PRIVATE: Done (excl. https://github.com/rust-bitcoin/rust-bitcoin/issues/6639)
- C-NEWTYPE-HIDE: Used `gg '\-\> .* {'` and read over the output.
- C-STRUCT-BOUNDS: Used `gg 'pub struct .*?<'`. No structs have trait bounds.

**Comment by tcharding** [comment]
`Debuggability` section

- C-DEBUG
- C-DEBUG-NONEMPTY

Done by: https://github.com/rust-bitcoin/rust-bitcoin/pull/6643

**Comment by mpbagot** [comment]
`Flexibility` section
 - C-INTERMEDIATE: Firstly, since #6412, `Target::from_compact` returns 0 for either a negative, or overflow case. Since there are multiple reasons for the default case 0 now, users have to do extra checks to figure out which, which implies that it might be worth instead making this fallible, or providing flags like Core does. Secondly, most error types have pub(crate) fields that are used for the Display strings. Should we provide getters for these to prevent users trying to parse Display strings?
 - C-CALLER-CONTROL: No `.clone()` use outside of tests.
 - C-GENERIC: Functionality seems generic outside of obvious constraints (array lengths, etc). No uses of slice parameters outside of tests and serde.
 - C-OBJECT: Only one public trait, `Integer`, which is explicitly not object safe because of the `FromStr` bound.

**Comment by mpbagot** [comment]
`Predictability` section
- C-SMART-PTR: No smart pointers in the crate.
- C-CONV-SPECIFIC: `Target` has a `from_compact`, but `CompactTarget` has no `to_target`. There is a `From<CompactTarget> for Target` though, which might be fine. I've added a `CompactTarget::to_target` in #6650
- C-METHOD: The only bare functions in the crate are those in parse int that operate on strings. Strings are foreign types so we can't make them methods. All good.
- C-NO-OUT: All good. The only methods with output `&mut` args are formatting and arbitrary impls.
- C-OVERLOAD: These mostly look correct. Return types seem correct and most are obvious wrappers around the underlying primitive type. Opened #6652 for `Rem<Weight>` which appears incorrect.
- C-DEREF: No `Deref`/`DerefMut` impls in the crate.
- C-CTOR: Constructors all look fine. `new`, where used, is obviously a primary method for constructing/computing a type. `from_` use is appropriate. Only thing to mention is the `from_height` methods on `absolute::Locktime`, `relative::LockTime`, `Sequence` and `NumberOfBlock`. One of these takes an absolute block height, the other a relative count of blocks. The docs are decent, but the names are sub-par. Should these be adjusted?

**Comment by tcharding** [comment]
>  The docs are decent, but the names are sub-par. Should these be adjusted?

Caveat: lock times a seriously fraught with danger. Everything about them is hard.

 I re-checked the `from_height` names and slightly agree that perhaps we could say 'block count' instead of 'block height' in docs on the relative locktime stuff. But not enough for me to do the patch. 

(FWIW I am of the personal opinion that locktimes in Bitcoin are unusable, even with all the API effort we put in, unless you _really_ understand them.)

**Comment by mpbagot** [comment]
`Dependability` section
- C-VALIDATE: `Target::from_compact`/`CompactTarget::to_target` are both infallible and coerce invalid (negative or overflow) values to `Target::ZERO`. This doesn't follow the spirit of C-VALIDATE, but would necessitate a new error type and change to the signature.
- C-DTOR-FAIL: No destructors in the crate.
- C-DTOR-BLOCK: No destructors in the crate.

**Comment by mpbagot** [comment]
`Type safety` section
- C-NEWTYPE: This is basically the purpose of this crate. After `Sequence` is closed, I don't see any problems here.
- C-CUSTOM-TYPE: Primitive stdlib types are only used for numbers in constructors or accessors. This is mostly fine. Some constructors like `absolute::LockTime::from_height/from_mtp` take a `u32`, since they're just a trivial wrapper if they take a concrete `Height`/`MedianTimePast` type.
- C-BITFLAG: The only type close to a flag set is `Sequence`, but it's only partly this, as it also encodes a number in the bottom bits. I think it's fine.
- C-BUILDER: There are no sufficiently complex types in `units` to warrant a builder.

**Comment by tcharding** [comment]
> and coerce invalid (negative or overflow) values to Target::ZERO. 

AFAIU this is ok because target of zero can never be achieved so this is functionally the same as an error case.

**Comment by tcharding** [comment]
Posting this so we can track the second pass. Just listing whats left as of this post.

- [ ] Naming** *(crate aligns with Rust naming conventions)
- [x] Interoperability** *(crate interacts nicely with other library functionality)
- [x] Macros** *(crate presents well-behaved macros)
- [ ] Documentation** *(crate is abundantly documented)
- [x] Predictability** *(crate enables legible code that acts how it looks)
- [x] Flexibility** *(crate supports diverse real-world use cases)
- [x] Type safety** *(crate leverages the type system effectively)
- [x] Dependability** *(crate is unlikely to do the wrong thing)
- [x] Debuggability** *(crate is conducive to easy debugging)
- [x] Future proofing** *(crate is free to improve without breaking users' code)
- [x] Necessities** *(to whom they matter, they really matter)

**Comment by tcharding** [comment]
`Macros` section - No macros in the crate

**Comment by mpbagot** [comment]
You already did `Macros`, `Future proofing` and `Necessities` above last week.

**Comment by mpbagot** [comment]
`Interoperability` section
- C-COMMON-TRAITS: `Weight`, `FeeRate`, `Target`, `Work`, `BlockTime` and `BlockHeight` are all missing `Default`. Do any of these need a `Default`? `NumOpResult` is missing `Hash`, `PartialOrd` and `Ord`. Adding them with derives requires adding them to `NumOpError` and thus `MathOp`, but would allow results to be ordered like their wrapped numbers. Do we want them, and if so, should they be implemented with derive, or manually to avoid implementing them on `NumOpError`/`MathOp`?
- C-CONV-TRAITS: No use of `Into`/`TryInto` in the crate.
- C-COLLECT: No collection types in the crate.
- C-SERDE: Every type has serde traits directly implemented except for `Amount`, `SignedAmount` and `FeeRate`. These types all have accompanying serde modules depending on how users want to de/serialize.
- C-SEND-SYNC: Every type has these according to the api files.
- C-GOOD-ERR: All errors have `std::error::Source + Send + Sync` impls, lowercase error messages with no ending punctuation. 
- C-NUM-FMT: All numeric wrapper types have `Octal`, `Binary` and `Lower/UpperHex`. The only exception is `FeeRate`, which is intentional per [this comment](https://github.com/rust-bitcoin/rust-bitcoin/issues/5586#issuecomment-3823875480)
- C-RW-VALUE: No generic Read/Write functions in the crate.

We should also double-check #6643 to make sure the api test will cover everything above. For example, `Structs` is missing `Target` and `Work`.

**Comment by tcharding** [comment]
I just pushed to #6643. Needs checking still because I'm not real sharp right this second.

**Comment by tcharding** [comment]
I _think_ common traits is right, I did think about it a bit last week. But not totally sure. The refactor in #6643 makes it clear at least what implements what.

**Comment by mpbagot** [comment]
`Naming` section
- C-CONV: `NumOpResult` is `Copy` but has an `into_result` conversion function. Being a `Copy` type, this is against convention (should be `to_result`). `Denomination::as_str` is owned -> borrowed for a `Copy` type. It returns a `&'static str`, so it functions like a free borrowed -> borrowed, so it's probably fine.
- C-GETTER: No `get_` getters in the crate. Getters are named correctly.
- C-ITER: No iterators in the crate.
- C-ITER-TY: No iterators in the crate.
- C-FEATURE: All feature names are single words without prefixes.
- C-WORD-ORDER: A few inconsistent names here:
  - `Amount/SignedAmount::from_int_btc` differs from the other constructors which are all `from_<unit>_<type>`. Should it be renamed?
  - `LockTime` and `Sequence` have `from_consensus` and `to_consensus_u32`, while `CompactTarget` has `from_consensus` and `to_consensus`. Should these be uniform, and which way?
  - `Weight` uses `from_vb*` but `to_vbytes_`. `FeeRate` uses `to/from_vb*`. Does this matter?

**Comment by tcharding** [comment]
Continued the naming post here: https://github.com/rust-bitcoin/rust-bitcoin/discussions/6673

**Comment by tcharding** [comment]
Docs section

- C-EXAMPLE: None needed, these are simple types with simple functions
- C-QUESTION-MARK: good, no uses of `unwrap` in docs (except one intentionally)
- C-FAILURE: We have a lint for this
- C-LINK: We have a lint for this
- C-METADATA: Good
- C-RELNOTES: Good
- C-HIDDEN: Pretty sure this is fine

**Comment by tcharding** [comment]
BOOM! The API checklist is done, again. Naming stuff is tracked in a discussion and on #3556. I'm closing this. Thanks @mpbagot for your grind work man, we couldn't get these 1.0's out without you.
